src/ecs/Engine.hx

```diff
@@ -337,4 +341,41 @@ class Engine {
 
         return target_system;
     }
+
+    public function addGlobalSystem<T:ecs.System>(systemClass:Class<T>) {
+        var target_system:System = getOrCreateSystem(systemClass);
+        globalSystems.push(target_system);
+        enable(systemClass);
+    }
+
+    public function setStateSystem<T:ecs.System>(state:Int, systemClass:Class<T>) {
+        var target_system:System = getOrCreateSystem(systemClass);
+
+        if(!stateSystems.exists(state)) {
+            stateSystems[state] = [];
+        }
+
+        stateSystems[state].push(target_system);
+    }
+
+    public function changeState(state:Int) {
+        var system:System = systemList.head;
+        var nextSystems = stateSystems[state];
+
+        if(nextSystems == null) {
+            throw "Unknown state " + state;
+        }
+
+        while(system != null) {
+            if(!globalSystems.contains(system) && !nextSystems.contains(system)) {
+                removeSystem(system);
+            }
+
+            system = system.next;
+        }
+
+        for(system in nextSystems) {
+            addSystem(system, system.priority);
+        }
```

Comment:
If a system was in previous states, it stays in the previous loop, but will be added again here. Is it safe?

Reply:
good catch

Reply:
done